crates/uv-auth/src/credentials.rs

```diff
@@ -1,117 +1,156 @@
             // See <https://github.com/pypa/pip/blob/06d21db4ff1ab69665c22a88718a4ea9757ca293/src/pip/_internal/utils/misc.py#L497-L499>
-            username: urlencoding::decode(url.username())
-                .expect("An encoded username should always decode")
-                .into_owned(),
+            username: if url.username().is_empty() {
+                None
+            } else {
+                Some(
+                    urlencoding::decode(url.username())
+                        .expect("An encoded username should always decode")
+                        .into_owned(),
+                )
+            },
             password: url.password().map(|password| {
                 urlencoding::decode(password)
                     .expect("An encoded password should always decode")
                     .into_owned()
             }),
         })
     }
-}
 
-impl From<Authenticator> for Credentials {
-    fn from(auth: Authenticator) -> Self {
-        Credentials {
-            username: auth.login,
-            password: Some(auth.password),
+    /// Parse [`Credentials`] from an HTTP request, if any.
+    ///
+    /// Only HTTP Basic Authentication is supported.
+    pub fn from_request(request: &Request) -> Option<Self> {
+        // First, attempt to retrieve the credentials from the URL
+        Self::from_url(request.url()).or(
+            // Then, attempt to pull the credentials from the headers
+            request
+                .headers()
+                .get(reqwest::header::AUTHORIZATION)
+                .map(Self::from_header_value)?,
+        )
+    }
+
+    /// Parse [`Credentials`] from an authorization header, if any.
+    ///
+    /// Only HTTP Basic Authentication is supported.
+    ///
+    /// [`None`] will be returned if any error is encountered.
+    pub(crate) fn from_header_value(header: &HeaderValue) -> Option<Self> {
+        let mut value = header.as_bytes().strip_prefix(b"Basic ")?;
+        let mut decoder = DecoderReader::new(&mut value, &BASE64_STANDARD);
+        let mut buf = String::new();
+        decoder.read_to_string(&mut buf).ok()?;
+        let (username, password) = buf.split_once(':')?;
+        let username = if username.is_empty() {
+            None
+        } else {
+            Some(username.to_string())
+        };
+        let password = if password.is_empty() {
+            None
+        } else {
+            Some(password.to_string())
+        };
+        Some(Self::new(username, password))
+    }
+
+    /// Create an HTTP Basic Authentication header for the credentials.
+    pub(crate) fn to_header_value(&self) -> HeaderValue {
+        // See: <https://github.com/seanmonstar/reqwest/blob/2c11ef000b151c2eebeed2c18a7b81042220c6b0/src/util.rs#L3>
+        let mut buf = b"Basic ".to_vec();
+        {
+            let mut encoder = EncoderWriter::new(&mut buf, &BASE64_STANDARD);
+            let _ = write!(encoder, "{}:", self.username().unwrap_or_default());
```

Comment:
Should we `.expect` these, like below?

Reply:
No it's valid for `username` to be `None` and default to an empty string.

Reply:
I meant `.expect` on the result of `write!`.

Reply:
Right now it's being dropped.

Reply:
But maybe that's fine? `write!` can return an error and `let _ = ...` just drops the error.

Reply:
Oh I see what you mean. It does seem weird to drop the error here, I think we to know if we failed to write the credentials. I guess I'll add `expect`.